### ISSUE TITLE
metro-plugin-typescript throws false errors

### ISSUE THREAD

**Issue by younes0** [microsoft/rnx-kit#2247]
### What happened?

metro-plugin-typescript throws unexpected false errors such as:
`error TS2305: Module '"react-native-screens"' has no exported member 'enableScreens'.`
tsc doesn't throws these errors nor does VS Code.

I'm using Expo 48:
```js
const { getDefaultConfig } = require("expo/metro-config");
const { TypeScriptPlugin } = require("@rnx-kit/metro-plugin-typescript");

module.exports = (() => {
  const config = getDefaultConfig(__dirname);
  config.serializer = {
    experimentalSerializerHook: TypeScriptPlugin(),
  };
  return config;
})();
```

### Affected Package

@rnx-kit/metro-plugin-typescript

### Version

0.3.3

### Which platforms are you seeing this issue on?

- [ ] Android
- [ ] iOS
- [ ] macOS
- [X] Windows

### System Information

```true
System:
    OS: Linux 5.15 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i9-12900H
    Memory: 11.06 GB / 15.62 GB
    Shell: 3.3.1 - /usr/bin/fish
  Binaries:
    Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
  IDEs:
    Android Studio: Not Found
  Languages:
    Java: 11.0.17 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.2.0 => 18.2.0
    react-native: 0.71.3 => 0.71.3
  npmGlobalPackages:
    *react-native*: Not Found
```

### Steps to Reproduce

add react-native-screens to your project and add
`import { enableScreens } from "react-native-screens";`

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by tido64** [comment]
Hi younes0, thanks for reporting this.

I've had a look and it looks like the error is both correct and wrong: According to `react-native-screens/package.json`, `types` points to `lib/typescript/index`. Looking inside the package, I found both an `index.d.ts` and `index.native.d.ts`. I don't know the story behind the latter alternative as platform forking doesn't apply to files pointed to by main fields, but our plugin incorrectly picks the latter where `enableScreens` indeed isn't declared.

In any case, I've changed the plugin to not add platform suffixes when resolving main fields. It would be great if you could also check that it works as expected.

**Comment by younes0** [comment]
@tido64 
Thanks for your response.

I should have post the entire errors logs: I have many other errors (with [react-native-paper](https://callstack.github.io/react-native-paper/)) that I had with 0.3.3, and still have with 0.3.4.

```
 Property 'theme' is missing in type '{ icon: string; size: number; onPress: (() => void) | undefined; style: ReactNativeStyle; }' but required in type 'Omit<Props, "ref">'.

26         <IconButton
            ~~~~~~~~~~

  node_modules/react-native-paper/lib/typescript/components/TouchableRipple/TouchableRipple.native.d.ts:13:5
    13     theme: InternalTheme;
           ~~~~~
    'theme' is declared here.
src/components/appbar/ModalAppbar.tsx:27:17 - error TS2741: Property 'theme' is missing in type '{ icon: string; onPress: () => void; }' but required in type 'Omit<Props, "ref">'.

27     {onClose ? <Appbar.Action icon={"close"} onPress={onClose} /> : null}
                   ~~~~~~~~~~~~~

  node_modules/react-native-paper/lib/typescript/components/TouchableRipple/TouchableRipple.native.d.ts:13:5
    13     theme: InternalTheme;
           ~~~~~
    'theme' is declared here.
src/components/appbar/CustomAppbar.tsx:34:10 - error TS2741: Property 'theme' is missing in type '{ onPress: () => void; }' but required in type 'Omit<Props, "ref">'.

34         <Appbar.BackAction onPress={() => navigation.goBack()} />
            ~~~~~~~~~~~~~~~~~

  node_modules/react-native-paper/lib/typescript/components/TouchableRipple/TouchableRipple.native.d.ts:13:5
    13     theme: InternalTheme;
           ~~~~~
    'theme' is declared here.
src/components/sheet/SheetHeader.tsx:26:10 - error TS2741: Property 'theme' is missing in type '{ icon: string; size: number; onPress: (() => void) | undefined; style: ReactNativeStyle; }' but required in type 'Omit<Props, "ref">'.

26         <IconButton
            ~~~~~~~~~~

  node_modules/react-native-paper/lib/typescript/components/TouchableRipple/TouchableRipple.native.d.ts:13:5
    13     theme: InternalTheme;
           ~~~~~
    'theme' is declared here.
src/components/appbar/ModalAppbar.tsx:27:17 - error TS2741: Property 'theme' is missing in type '{ icon: string; onPress: () => void; }' but required in type 'Omit<Props, "ref">'.

27     {onClose ? <Appbar.Action icon={"close"} onPress={onClose} /> : null}
                   ~~~~~~~~~~~~~

  node_modules/react-native-paper/lib/typescript/components/TouchableRipple/TouchableRipple.native.d.ts:13:5
    13     theme: InternalTheme;
           ~~~~~
    'theme' is declared here.
```

**Comment by tido64** [comment]
I'm not familiar with `react-native-paper` but I downloaded the package and looked at how `IconButton` and `TouchableRipple` are declared, and it looks like those errors are correct. I made a minimal repro here: [TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKgFhAttAvFA3gXwNwCheiSwD2ArgMZwCGARgDYQBKAlmGA1GurlL1MAmQAuWIIh5M+QtACS5YgDsAQqWDBFnEhWr0mrdtABkGHnwFIIAfhHwLE-PIUBnYFBqr1CkXMUq1GrhxTKAB6EKgAPVhwaAByLFioZicoBWJXKicnZgBzBVoOdX4YqFifZQ9FWIA6YLC+AAUAJ2JIJtBS82RE5KhEZOyFHKSFYqJ4zET3VyaIAEdSZlmAExGxuJgySgK9NgYagAoAJgBmI6OASlwgA)

![image](https://user-images.githubusercontent.com/4123478/221951426-ef2a66d3-f0a7-40b6-bb18-9e6ca00d018c.png)

This might be something you should file with the `react-native-paper` folks.

**Comment by younes0** [comment]
@tido64

but _tsc_ and _VS Code_ do not raise these errors.
Here is my tsconfig.json if it can help:

```json
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "baseUrl": "./",
    "declaration": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": false,
    "incremental": true,
    "module": "commonjs",
    "noFallthroughCasesInSwitch": false,
    "noImplicitAny": false,
    "outDir": "./dist",
    "removeComments": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strictBindCallApply": false,
    "strictNullChecks": false,
    "target": "es2017",
    "types": ["node", "jest"]
  }
}
```

**Comment by younes0** [comment]
Using "typescript": "4.9.4",

**Comment by tido64** [comment]
Okay, I think I've figured this out. The reason it's different is because we inject [`moduleSuffixes`](https://www.typescriptlang.org/tsconfig#moduleSuffixes) if it's not set in `tsconfig.json`. This setting makes TypeScript aware of platform suffixes like `.android`, `.ios`, `.native`. When you run `tsc` with the current config, it just resolves files normally without considering platform forked files.

In this specific scenario, `react-native-paper` has two variants of `TouchableRipple`: [`TouchableRipple.tsx`](https://github.com/callstack/react-native-paper/blob/e9f5dafe93348d8a72c931bc9e56afa6d2769b45/src/components/TouchableRipple/TouchableRipple.tsx#L57) and [`TouchableRipple.native.tsx`](https://github.com/callstack/react-native-paper/blob/e9f5dafe93348d8a72c931bc9e56afa6d2769b45/src/components/TouchableRipple/TouchableRipple.native.tsx#L28). `tsc` will pick `.tsx`, but our plugin will pick `.native.tsx`. If you open these up, you'll see that the `Props` type differ.

```ts
// TouchableRipple.tsx
export type Props = React.ComponentPropsWithRef<typeof Pressable> & {
  ...
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  theme?: ThemeProp;
};

// TouchableRipple.native.tsx
type Props = React.ComponentProps<typeof Pressable> & {
  ...
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  theme: InternalTheme;
};
```

In `.tsx`, the `theme` prop is optional. Whereas in `.native.tsx`, it's required. So the error you're getting here is actually more accurate than what `tsc` provides with the current config.

If you add the following line to your `tsconfig.json`, you will also see the error with `tsc`:

```diff
     "target": "es2017",
     "types": ["node", "jest"],
+    "moduleSuffixes": [".android", ".native", ""]
   }
 }
```

Here's the output from my local testing:

```
% yarn tsc --outDir lib
src/App.native.tsx:196:12 - error TS2741: Property 'theme' is missing in type '{ icon: string; size: number; onPress: () => void; }' but required in type 'Pick<Props, "testID" | "disabled" | "hitSlop" | "onLayout" | "pointerEvents" | "removeClippedSubviews" | "style" | "nativeID" | "collapsable" | "needsOffscreenAlphaCompositing" | ... 72 more ... | "size">'.

196           <IconButton
               ~~~~~~~~~~

  ../../node_modules/react-native-paper/lib/typescript/components/TouchableRipple/TouchableRipple.native.d.ts:13:5
    13     theme: InternalTheme;
           ~~~~~
    'theme' is declared here.


Found 1 error in src/App.native.tsx:196
```

In short, this is indeed a bug in `react-native-paper`. You should file an issue there with this information.

As for workarounds, the plugin currently doesn't allow you to disable this. And you probably wouldn't want to either as it's uncovering bugs as intended. I would suggest that you use [patch-package](https://www.npmjs.com/package/patch-package) to make that field optional. I've looked at how it gets used and it looks like it is assumed to be optional regardless so it should fine.

**Comment by younes0** [comment]
@tido64 got it, thanks for the explanations. I will open an issue on `react-native-paper`
